File: examples/ch06/py/before.py
"""Chapter 6 · BEFORE: a cart that cannot protect itself.

Everything is public, so nothing is safe. `total` is a plain attribute the
caller must remember to recompute (they won't), and `discount_rate` accepts
any number at all. The object has no way to keep its own promises.
"""

from dataclasses import dataclass, field


@dataclass(frozen=True)
class LineItem:
    name: str
    price: float


@dataclass
class ShoppingCart:
    items: list[LineItem] = field(default_factory=list)
    discount_rate: float = 0.0
    total: float = 0.0  # a cached field the caller must keep in sync by hand

    def recompute(self) -> None:
        subtotal = sum(item.price for item in self.items)
        self.total = round(subtotal * (1 - self.discount_rate), 2)

File: examples/ch06/ts/before.ts
// Chapter 6 · BEFORE: a cart that cannot protect itself.
//
// Everything is a public mutable field, so nothing is safe. `total` is a plain
// number the caller must remember to recompute (they won't), and `discountRate`
// accepts any number at all. The object has no way to keep its own promises.

import { type LineItem } from "./models";

export class ShoppingCart {
  items: LineItem[] = [];
  discountRate = 0; // no guard: 2.0 (200% off) is accepted without complaint
  total = 0; // a cached field the caller must keep in sync by hand

  recompute(): void {
    const subtotal = this.items.reduce((sum, item) => sum + item.price, 0);
    this.total = Math.round(subtotal * (1 - this.discountRate) * 100) / 100;
  }
}
